### DISCUSSION TITLE
Understanding Reactivity

### DISCUSSION THREAD

**Discussion by jcuenod** [solidjs/solid#1215] [Q&A]
Hi, I've been following solid for some time, but I'm wondering about a design decision.

In the example below, I would expect `setMyArray` to trigger a re-render of `myArray` in the return value. I suppose that it's because behind the scenes, setting `myNewArray = myArray()` sets `myNewArray` to point to `myArray()` and so the old and new values look equal. Obviously this is not the case, though, and `myArray()` does return the expected output when it's subsequently called. Why not do a deeper equality check (assuming I'm understanding the internals correctly)?

```
import { render } from "solid-js/web";
import { createSignal } from "solid-js";

function Counter() {
  const [myArray, setMyArray] = createSignal([0]);
  const increment = () => {
    const newNum = Math.round(Math.random() * 100)
    let myNewArray = myArray()
    myNewArray.push(newNum)
    console.log(myArray(), myNewArray)
    setMyArray(myNewArray)
  };

  return <>
    <button type="button" onClick={increment}>
      click
    </button>
    <br />
    {JSON.stringify(myArray())}
  </>;
}

render(() => <Counter />, document.getElementById("app")!);
```

**Comment by atk** [comment]
The signal compares the new and old value by reference. The new array reused the reference from the old one and is therefore ignored by solid's reactivity.

You can either shallow-clone the array or use `{ equals: false }` in the signal's options to avoid this.

**Reply by jcuenod** [reply]
Ahh, thank you. Yes, as I said, I understand the fact that `myNewArray` is pointer, but I wasn't aware of the `equals` option (or that it can be set to a [comparison function](https://www.solidjs.com/docs/latest/api#options)).

```
const [myString, setMyString] = createSignal("string", {
  equals: (newVal, oldVal) => newVal.length === oldVal.length,
});
```

But this still seems like an obvious gotcha to me. So I'm wondering what the reason for the design decision is. I guess that @ryansolid benchmarked the tradeoff, but I'd be curious to know what that tradeoff was... (like does solid then run at react speeds on the js benchmarks?)

**Reply by ryansolid** [reply]
Equals false has no performance cost, it's just a choice. We used to be equals false but it created inconsistency because most of the system does do these comparisons. Things like derived data propagation. Best to just start out right.

But doing deep diffs on data is a performance hit felt everywhere. And yeah if you were to deep diff everything, you'd be going back into VDOM territory. Actually possibly worse because VDOM data is immutable so you can rely on structural cloning most of the time to shortcut diffs where we couldn't.

Some reactive systems only do equality checks on primitive data but that also feels inconsistent. I think it is best to treat things immutable at the source of change. It's also consistent with the rest of our APIs like `createStore`.
